File: JavaScript/cp/CountAnagrams.js
/*  Given two strings word and text, return the number of 
substrings where text contains any anagram of word. */


/* Function to count the number of anagrams */
const countAnagrams = (word, text) => {
    let alpha = new Array(26);

    for(let i = 0; i<alpha.length; i++){
        alpha[i] = 0;
    }

    /* Calculating the frequency of each character in the given word */
    for(let c in word){
        alpha[word.charCodeAt(c)-97]++;
    }

    let window = [];
    let ans = 0;

    /* Checking the occurrences of anagrams of the given word*/
    for(let i=0; i<text.length; i++){
        let c = text.charCodeAt(i);

        /* Adding the current character to the window if the word contains it*/
        if(alpha[c-97] > 0){
            /* Subtracting the given word as we have added it to the window */
            alpha[c-97]--;
            window.push(i);
        }else{

            /* Decrementing the window until it is feasible to accommodate the current character */
            while(window.length > 0 && alpha[c-97] <= 0){
                alpha[text.charCodeAt(window.shift()) - 97]++;
            }

            /* Adding the current character to the window if the word contains it */
            if(alpha[c-97] > 0){
                alpha[c-97]--;
                window.push(i);
            }
        }

        /* Increasing the answer if the window size = word size */
        if(window.length == word.length){
            ans++;
            /* Removing the first character of the window */
            alpha[text.charCodeAt(window.shift()) - 97]++;
        }
    }

    return ans;
}


console.log("Enter the word to be found and the text in which the word needs to be searched");

/* List stores the word and the text */
let list = readline()
    .split(" ")
    .map((element) => element);

/* Only two elements are present in the list, the first one 
represents the word to be searched and the second one 
represents the text in which the word is searched in  */
console.log("Number of substrings which contains the anagram ", list[0], " in ", list[1], " is ", countAnagrams(list[0], list[1]));

/* 
Example 1:

Input:

ab aab

Output:

Number of substrings which contains the anagram ab in aab is 1

Time Complexity: O(N)
Space Complexity: O(K)
*/

File: Python/cp/CountAnagrams.py
'''
Given two strings <target> and <text>, write a program to find the number
of anagrams of <target> that can be found in the string <text>
Length of the <target> <= length of <text> <= 10^5
'''

from collections import Counter


def countAnagrams(target, text):

    '''
    This function returns the number of anagrams of string <target> which are
    present in the string <text>

    Parameters:
    target: string whose anagrams need to be counted
    text: string where the anagrams need to be counted

    '''

    k = len(target)
    n = len(text)
    # initialize a hashmap/ dictionary which stores the key value pairs
    # for frequencies of all the characters of the target string
    required = Counter(target)
    # similarly initialize the window with the first k characters
    # of the string <text>

    window = Counter(text[:k])
    # this variable will contain the number of anagrams
    answer = 0
    for i in range(n-k):
        # checking if all the characters of the current window are matching
        # with the characters of the target
        if window & required == required:
            answer += 1
        # decrementing the frequency of the character outside the window
        # incrementing the frequency of the character entering the window
        window[text[i]] -= 1
        window[text[i+k]] += 1
    # computing the answer for the last window
    if window & required == required:
        answer += 1
    return answer


# Driver Code
if __name__ == "__main__":

    s1 = input('Enter the target string : ')
    s2 = input('Enter the text string : ')

    print('The number of anagrams of "{}" in "{}" are: '.format(s1, s2))
    print(countAnagrams(s1, s2))

problemDetails = '''
Time Complexity:
O(n) where n is the length of the string <text>
additional time complexity : O(k) where k is length of the target string
since n > k : overall time complexity of the code => O(n)

Space complexity:
O(1) is the overall space complexity of this code, since the length of the
window cannot excede 26 ( alphabet)

Test cases:
TC-1
Input:
Enter the target string : abc
Enter the text string : bcabxabc
Output:
The number of anagrams of "abc" in "bcabxabc" are:  3

TC-2
Input:
Enter the target string : algo
Enter the text string : neoalgoneoprogalgoolagneoprogalog
Output"
The number of anagrams of "algo" in "neoalgoneoprogalgoolagneoprogalog" are:  8
'''
